QSPIFBlockDevice.h

```diff
@@ -0,0 +1,274 @@
+
+    // Send Bus configure_format command to Driver
+    qspi_status_t _qspi_configure_format(qspi_bus_width_t inst_width, qspi_bus_width_t address_width,
+                                      qspi_address_size_t address_size, qspi_bus_width_t alt_width, qspi_alt_size_t alt_size, qspi_bus_width_t data_width,
+                                      int dummy_cycles);
+
+    // Send set_frequency command to Driver
+    qspi_status_t _qsp_set_frequency(int freq);
+    /********************************/
+
+
+    // Soft Reset Flash Memory
+    int _reset_flash_mem();
+
+    // Configure Write Enable in Status Register
+    int _set_write_enable();
+
+    // Wait on status register until write not-in-progress
+    bool _is_mem_ready();
+
+
+    /* SFDP Detection and Parsing Functions */
+    /****************************************/
+    // Parse SFDP Headers and retrieve Basic Param and Sector Map Tables (if exist)
+    int _sfdp_parse_sfdp_headers(uint32_t& basic_table_addr, size_t& basic_table_size,
+                              uint32_t& sector_map_table_addr, size_t& sector_map_table_size);
+
+    // Parse and Detect required Basic Parameters from Table
+    int _sfdp_parse_basic_param_table(uint32_t basic_table_addr, size_t basic_table_size);
+
+    // Parse and read information required by Regions Secotr Map
+    int _sfdp_parse_sector_map_table(uint32_t sector_map_table_addr, size_t sector_map_table_size);
+
+    // Detect fastest read Bus mode supported by device
+    int _sfdp_detect_best_bus_read_mode(uint8_t *basic_param_table_ptr, bool& set_quad_enable, bool& is_qpi_mode,
+                                   unsigned int& read_inst);
+
+    // Enable Quad mode if supported (1-1-4, 1-4-4, 4-4-4 bus modes)
+    int _sfdp_set_quad_enabled(uint8_t *basic_param_table_ptr);
+
+    // Enable QPI mode (4-4-4) is supported
+    int _sfdp_set_qpi_enabled(uint8_t *basic_param_table_ptr);
+
+    // Set Page size for program
+    int _sfdp_detect_page_size(uint8_t *basic_param_table_ptr);
+
+    // Detect all supported erase types
+    int _sfdp_detect_erase_types_inst_and_size(uint8_t *basic_param_table_ptr, unsigned int& erase4k_inst,
+                                         unsigned int *erase_type_inst_arr, unsigned int *erase_type_size_arr);
+
+    /* Utilities Functions */
+    /***********************/
+    // Find the region to which the given offset belong to
+    int _utils_find_addr_region(int offset);
+
+    // Iterate on all supported Erase Types of the Region to which the offset belong to.
+    // Iterates from highest type to lowest
+    int _utils_iterate_next_largest_erase_type(uint8_t& bitfield, int size, int offset, int boundry);
+
+private:
+    // Internal Members
+
+    // QSPI Driver Object
+    QSPI _qspi;
+
+    bool _is_initialized;
+
+    // mutex is required for sequence of driver commands that can not be disrupted
+    static SingletonPtr<PlatformMutex> _mutex;
```

Comment:
Using a singleton mutex across all instances means you cannot use two QSPI flash chips concurrently. Shouldn't "transactions" be a requirement on the driver? (I imagine something like `_qspi.lock()`)

Reply:
_qspi.lock is implemented in driver level to protect single driver operations: e.g. write .
BD lock is protecting from a sequence of operations that must be carried sequentially: e.g. (a) _set_write_enable , (b) write, (c) _is_mem_ready . The QSPI driver requires these 3 operations to be carried with no interruption.

Reply:
@donatieng - after further discussions, the BD mutex was moved to instance level to support multiple QSPI flash chips, as you suggested - Thanks!

Reply:
It looks like the QSPI driver itself has exposed lock/unlock functions:
https://github.com/ARMmbed/mbed-os/blob/feature-qspi/drivers/QSPI.h#L174-L176

Should we actually be using those?

Reply:
@geky - QSPI's lock is protecting the bus driver. QSPIF BD lock is protecting the Flash device: 
e.g.

QSPI's lock will lock 3 times, each of: Set Write Enable, Send Write Command, Register Read
QSPIF BD lock will lock "program" command for that flash device (on other threads) that contains all of the above 3 because another flash command, such as read, can not be handled in between (a second QSPI Flash device can use the bus driver in between the first flash device commands though)